lib/src/base.dart

```diff
@@ -511,6 +522,474 @@ class Event {
+  /// This can either be the browser this event ocurred in, or the user
+  /// agent of a web request that triggered the event.
+  final Browser browser;
+
+  const Contexts({
+    this.device,
+    this.operatingSystem,
+    this.runtimes,
+    this.app,
+    this.browser,
+  });
+
+  /// Produces a [Map] that can be serialized to JSON.
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> json = {};
+
+    Map<String, dynamic> deviceMap;
+    if (device != null && (deviceMap = device.toJson()).isNotEmpty) {
+      json['device'] = deviceMap;
+    }
+
+    Map<String, dynamic> osMap;
+    if (operatingSystem != null &&
+        (osMap = operatingSystem.toJson()).isNotEmpty) {
+      json['os'] = osMap;
+    }
+
+    Map<String, dynamic> appMap;
+    if (app != null && (appMap = app.toJson()).isNotEmpty) {
+      json['app'] = appMap;
+    }
+
+    Map<String, dynamic> browserMap;
+    if (browser != null && (browserMap = browser.toJson()).isNotEmpty) {
+      json['browser'] = browserMap;
+    }
+
+    if (runtimes != null) {
+      if (runtimes.length == 1) {
+        final Runtime runtime = runtimes[0];
+        if (runtime != null) {
+          final String key = runtime.key ?? 'runtime';
+          json[key] = runtime.toJson();
+        }
+      } else if (runtimes.length > 1) {
+        for (final runtime in runtimes) {
+          if (runtime != null) {
+            String key = runtime.key ?? runtime.name.toLowerCase();
+
+            if (json.containsKey(key)) {
+              int k = 0;
+              while (json.containsKey(key)) {
+                key = '$key$k';
+                k++;
+              }
+            }
+
+            json[key] = runtime.toJson()..addAll({"type": "runtime"});
+          }
+        }
+      }
+    }
+
+    return json;
+  }
+}
+
+/// This describes the device that caused the event.
+class Device {
+  /// The name of the device. This is typically a hostname.
+  final String name;
+
+  /// The family of the device.
+  ///
+  /// This is normally the common part of model names across generations.
+  /// For instance `iPhone` would be a reasonable family,
+  /// so would be `Samsung Galaxy`.
+  final String family;
+
+  /// The model name. This for instance can be `Samsung Galaxy S3`.
+  final String model;
+
+  /// An internal hardware revision to identify the device exactly.
+  final String modelId;
+
+  /// The CPU architecture.
+  final String arch;
+
+  /// If the device has a battery, this can be an floating point value
+  /// defining the battery level (in the range 0-100).
+  final double batteryLevel;
+
+  /// Defines the orientation of a device.
+  final Orientation orientation;
+
+  /// The manufacturer of the device.
+  final String manufacturer;
+
+  /// The brand of the device.
+  final String brand;
+
+  /// The screen resolution. (e.g.: `800x600`, `3040x1444`).
+  final String screenResolution;
+
+  /// A floating point denoting the screen density.
+  final String screenDensity;
```

Comment:
This has to be a `double`. I get the following error from Sentry:

There was 1 error encountered while processing this event
contexts.device.screen_density: Discarded invalid value

Reason | expected a floating point number
-- | --
Value | 2.625

Reply:
Good point. [Just reviewed the docs](https://docs.sentry.io/development/sdk-dev/event-payloads/contexts/#device-context) and it says:

>  Optional. A floating point denoting the screen density.

I'll review the types and open a PR to fix this

Reply:
Thanks! Also filed https://github.com/flutter/flutter/issues/53521.